src/metabase/models/permissions.clj

```diff
@@ -1292,20 +1299,34 @@
               (log/trace "Deleting block permissions entries for Group %d for Database %d" group-id db-id)
               (t2/delete! Permissions :group_id group-id, :object (database-block-perms-path db-id)))]
       (condp = schemas
-        :all (do
-               (revoke-db-schema-permissions! group-id db-id)
-               (delete-block-perms-for-this-db!)
-               (grant-permissions-for-all-schemas! group-id db-id))
-        :none  (do
-                 (revoke-db-schema-permissions! group-id db-id)
-                 (delete-block-perms-for-this-db!))
+        :all
+        (do
+          (revoke-db-schema-permissions! group-id db-id)
+          (delete-block-perms-for-this-db!)
+          (grant-permissions-for-all-schemas! group-id db-id))
+
+        :none
+        (do
+          (revoke-db-schema-permissions! group-id db-id)
+          (delete-block-perms-for-this-db!))
+
+        ;; Groups using connection impersonation for a DB should be treated the same as if they had full self-service
+        ;; data access.
+        :impersonated
+        (do
+          (revoke-db-schema-permissions! group-id db-id)
+          (delete-block-perms-for-this-db!)
+          (grant-permissions-for-all-schemas! group-id db-id))
```

Comment:
Your comment makes me wonder if [these](https://github.com/metabase/metabase/pull/30714/files?diff=split&w=1#diff-80daa6af38b41b4ef83c066c9e150953e0aebd0edf9bc1d47a46a4165ccbab49R1304-R1306) operations should be in a function so they don't get out of sync